// @tolgee/cli/dist/client/getApiKeyInformation.js
import { API_KEY_PAK_PREFIX } from './../constants.js';
import { handleLoadableError } from './TolgeeClient.js';
import { exitWithError } from './../utils/logger.js';
export const getApiKeyInformation = async (client, key) => {
    if (key.startsWith(API_KEY_PAK_PREFIX)) {
        const loadable = await client.GET('/v2/api-keys/current');
        if (loadable.response.status === 401) {
            exitWithError("Couldn't log in: the API key you provided is invalid.");
        }
        handleLoadableError(loadable);
        const info = loadable.data;
        const username = info.userFullName || info.username || '<unknown user>';
        return {
            type: 'PAK',
            key: key,
            username: username,
            project: {
                id: info.projectId,
                name: info.projectName,
            },
            expires: info.expiresAt ?? 0,
        };
    }
    else {
        const loadable = await client.GET('/v2/pats/current');
        if (loadable.response.status === 401) {
            exitWithError("Couldn't log in: the API key you provided is invalid.");
        }
        handleLoadableError(loadable);
        const info = loadable.data;
        const username = info.user.name || info.user.username;
        return {
            type: 'PAT',
            key: key,
            username: username,
            expires: info.expiresAt ?? 0,
        };
    }
};
