// @tolgee/cli/dist/client/ImportClient.js
import { pathToPosix } from '../utils/pathToPosix.js';
export const createImportClient = ({ apiClient }) => {
    return {
        async import(data) {
            const body = new FormData();
            for (const file of data.files) {
                // converting paths to posix style, so it's correctly matched on the server
                body.append('files', new Blob([file.data]), pathToPosix(file.name));
            }
            data.params.fileMappings = data.params.fileMappings.map((i) => ({
                ...i,
                // converting paths to posix style, so it's correctly matched on the server
                fileName: pathToPosix(i.fileName),
            }));
            body.append('params', JSON.stringify(data.params));
            return apiClient.POST('/v2/projects/{projectId}/single-step-import', {
                params: { path: { projectId: apiClient.getProjectId() } },
                body: body,
                bodySerializer: (r) => r,
            });
        },
    };
};
